### PR TITLE
fix Spyglass prowjob endpoint

### LINKED ISSUE #9502: Spyglass doesn't handle Prow jobs
1. The current deployment of Spyglass doesn't handle Prow jobs (via /view/prowjob/), which is a regression as it did work in some previous iteration. This seems to be because there's no ArtifactFetcher corresponding to Prow jobs.

2. After the /view/prowjob/ endpoint is fixed, the Spyglass link on Deck needs to be updated as well (I mistakenly hard coded it to /view/gcs/).

/cc @paulangton @krzyzacy @BenTheElder

### PR DESCRIPTION
**Pull request by ibzib**
fixes #9502
/cc @AishSundar @krzyzacy @BenTheElder @stevekuznetsov @paulangton 

Logs for pending jobs can now be viewed in Spyglass.
screen shot 2018-09-29 at 3 35 56 pm

In order to do this and make future changes easier, I made the following changes:
 
- implement PodLogArtifactFetcher (before, this logic was floating around in spyglass.go)
- consolidate Spyglass HTTP handlers

(Kind of unrelated) I also noticed the GCS artifact fetcher hanging, so I capped it at 500 tries per artifact before giving up.

### CODE REVIEW

**Comment by krzyzacy** [comment]
(to check if outstanding comments from offline discussion has resolved or not)

File: prow/spyglass/gcsartifact_fetcher.go [hunk 1]
```diff
@@ -62,15 +68,19 @@ func NewGCSArtifactFetcher(c *storage.Client) *GCSArtifactFetcher {
...
 		"jobPrefix": src.jobPath(),
 	}
 }
 
 // newGCSJobSource creates a new gcsJobSource from a given bucket and jobPrefix
 func newGCSJobSource(src string) (*gcsJobSource, error) {
-	gcsURL, err := url.Parse(src)
+	split := strings.SplitN(src, "/", 2)
+	if len(split) <= 1 {
+		return &gcsJobSource{}, ErrCannotParseSource
```
Comment (stevekuznetsov) [inline prow/spyglass/gcsartifact_fetcher.go]: This error is super opaque -- can we explain why the parsing failed for end user UX?

File: prow/spyglass/gcsartifact_fetcher.go [hunk 2]
```diff
@@ -97,31 +107,41 @@ func newGCSJobSource(src string) (*gcsJobSource, error) {
...
+		artifactNames = append(artifactNames, strings.TrimPrefix(oAttrs.Name, prefix))
 
 	}
+	if !done {
+		return artifactNames, fmt.Errorf("Gave up after %d tries", maxTries)
+	}
+
 	listElapsed := time.Since(listStart)
-	logrus.WithField("duration", listElapsed).Infof("Listed %d artifacts.", len(artifacts))
-	return artifacts
```
Comment (stevekuznetsov) [inline prow/spyglass/gcsartifact_fetcher.go]: Can you try to split out rote re-names in your commits? It's really hard to review the substance of this PR when so many diff lines are `s/oldName/newName/g`

File: prow/spyglass/gcsartifact_fetcher_test.go [hunk 3]
```diff
@@ -48,7 +51,7 @@ func TestNewGCSJobSource(t *testing.T) {
 		},
 		{
 			name:        "Test GCS link with org name",
-			src:         "gs://test-bucket/logs/sig-flexing/example-ci-run/403",
+			srcPath:     "gcs/test-bucket/logs/sig-flexing/example-ci-run/403",
```
Comment (stevekuznetsov) [inline prow/spyglass/gcsartifact_fetcher_test.go]: Why is this a useful change?

File: prow/spyglass/podlogartifact.go [hunk 4]
```diff
@@ -112,7 +113,7 @@ func (a *PodLogArtifact) ReadAll() ([]byte, error) {
 	if size > a.sizeLimit {
 		return nil, viewers.ErrFileTooLarge
 	}
-	logs, err := a.ja.GetJobLog(a.name, a.buildID)
+	logs, err := a.ja.GetJobLogByUUID(a.prowID)
```
Comment (stevekuznetsov) [inline prow/spyglass/podlogartifact.go]: Why is this a more appropriate lookup?

File: prow/spyglass/podlogartifact_fetcher.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+// NewPodLogArtifactFetcher returns a PodLogArtifactFetcher using the given job agent as storage
+func NewPodLogArtifactFetcher(ja podLogJobAgent) *PodLogArtifactFetcher {
+	return &PodLogArtifactFetcher{jobAgent: ja}
+}
+
+func (af *PodLogArtifactFetcher) artifact(src string, sizeLimit int64) (viewers.Artifact, error) {
+	split := strings.SplitN(src, "/", 2)
+	if len(split) < 2 {
+		return nil, fmt.Errorf("Invalid prow src: %s", src)
```
Comment (stevekuznetsov) [inline prow/spyglass/podlogartifact_fetcher.go]: The "why" of this is going to be very important -- error messages for end users need to help them achieve a correct request

File: prow/spyglass/spyglass.go [hunk 6]
```diff
@@ -48,13 +41,11 @@ type Spyglass struct {
 	// Lenses is a map of names of views to their corresponding lenses
 	Lenses map[string]Lens
 
-	// Eyepieces is a list of ArtifactFetchers used by Spyglass. Whenever spyglass
-	// receives a request to view artifacts from a source, it will check to see if it
-	// can handle that source with one of its registered eyepieces
-	Eyepieces []ArtifactFetcher
```
Comment (stevekuznetsov) [inline prow/spyglass/spyglass.go]: This seems like a large architectural change?

**Comment by BenTheElder** [comment]
Agree with the above. The PR body should elaborate further 👍

**Comment by ibzib** [comment]
Thanks for the feedback Steve. I'm going to split this into multiple PRs.

**Pull request closed without merging**